worker/pubsub/remoteserver.go

```diff
@@ -0,0 +1,336 @@
+		MaxDelay:    5 * time.Minute,
+		BackoffFunc: retry.DoubleDelay,
+		Stop:        stop,
+		Clock:       r.clock,
+	})
+
+	r.mutex.Lock()
+	r.stopConnecting = nil
+	r.mutex.Unlock()
+
+	if connection != nil {
+		logger.Debugf("forwarding connected %s -> %s", r.origin, r.target)
+		_, err := r.hub.Publish(
+			forwarder.ConnectedTopic,
+			// NOTE: origin is filled in by the the central hub annotations.
+			forwarder.OriginTarget{Target: r.target})
+		if err != nil {
+			logger.Errorf("%v", err)
+		}
+	}
+	return connection
+}
+
+func (r *remoteServer) loop() error {
+	defer r.unsubscribe()
+
+	var next <-chan struct{}
+	var delay <-chan time.Time
+	for {
+		if r.connection == nil {
+			// If we don't have a current connection, try to get one.
+			if connection := r.connect(); connection != nil {
+				r.mutex.Lock()
+				r.connection = connection
+				r.mutex.Unlock()
+				delay = nil
+			} else {
+				// Skip through the select to try to reconnect.
+				delay = r.clock.After(time.Second)
+			}
+		}
+
+		select {
+		case <-r.tomb.Dying():
+			r.mutex.Lock()
+			if r.connection != nil {
+				r.connection.Close()
+			}
+			r.mutex.Unlock()
+			return tomb.ErrDying
+		case <-r.data:
+			// Has new data been pushed on?
+		case <-next:
+			// If there was already data, next is a closed channel.
+			// otherwise it is nil so won't pass through.
+		case <-delay:
+			// If we failed to connect for whatever reason, this means we don't cycle
+			// immediately.
+		}
+
+		// If we are not currently connected, no point trying to pop off any
+		// messages that aren't there.
+		if r.connection == nil {
+			continue
+		}
+
+		message, empty := r.nextMessage()
+		if empty {
+			next = nil
+		} else {
+			next = r.closed
+		}
+		if message != nil {
+			r.forwardMessage(message)
+		}
+	}
+}
+
+func (r *remoteServer) forwardMessage(message *params.PubSubMessage) {
+	logger.Tracef("forwarding %q to %s, data %v", message.Topic, r.target, message.Data)
+	err := r.connection.ForwardMessage(message)
+	if err != nil {
+		// Some problem sending, so log, close the connection, and try to reconnect.
+		logger.Infof("unable to forward message, reconnecting... : %v", err)
+		r.mutex.Lock()
+		defer r.mutex.Unlock()
+		r.connection.Close()
+		r.connection = nil
+		// Discard all pending messages.
+		r.pending = deque.New()
+		// Tell everyone what we have been disconnected.
+		_, err = r.hub.Publish(
+			forwarder.DisconnectedTopic,
+			// NOTE: origin is filled in by the the central hub annotations.
+			forwarder.OriginTarget{Target: r.target})
+		if err != nil {
+			logger.Errorf("%v", err)
+		}
+	}
+}
+
+func (r *remoteServer) interruptConnecting() {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+	if r.stopConnecting != nil {
+		logger.Debugf("interrupting the pending connect loop")
+		close(r.stopConnecting)
+		r.stopConnecting = nil
+	}
+}
+
+// Kill is part of the worker.Worker interface.
+func (r *remoteServer) Kill() {
+	r.tomb.Kill(nil)
+	r.interruptConnecting()
```

Comment:
Well if you think about it, there are only two ways, either it is killed from the outside, or the loop function returns an error. If the loop function returns an error, there is nothing to interrupt.